packages/server/src/lib/grants-ingest.js

```diff
@@ -0,0 +1,156 @@
+        open_date: normalizeDateString(messageData.PostDate),
+        close_date: normalizeDateString(messageData.CloseDate || '2100-01-01'),
+        notes: 'auto-inserted by script',
+        search_terms: '[in title/desc]+',
+        reviewer_name: 'none',
+        opportunity_category: opportunityCategoryMap[messageData.OpportunityCategory],
+        description: messageData.Description,
+        eligibility_codes: (messageData.EligibleApplicants || []).join(' '),
+        opportunity_status: 'posted',
+        raw_body: JSON.stringify(messageData),
+    };
+}
+
+async function upsertGrant(knex, grant) {
+    await knex('grants')
+        .insert(grant)
+        .onConflict('grant_id')
+        .merge({ ...grant, ...{ updated_at: 'now' } })
+        .returning('grant_id');
+}
+
+async function deleteMessage(sqs, queueUrl, receiptHandle) {
+    const command = new DeleteMessageCommand({
+        QueueUrl: queueUrl,
+        ReceiptHandle: receiptHandle,
+    });
+    await sqs.send(command);
+}
+
+/**
+ * processMessages Saves a batch of SQS messages containing JSON grant data sent
+ * from the grants-ingest service to the database. Existing database records that match
+ * a message's grant identifier are updated. After each message is processed, it is
+ * deleted from the SQS queue.
+ *
+ * Any errors related to parsing or saving are logged and do not prevent further processing.
+ * Errors interacting with SQS are fatal.
+ *
+ * @param { import('knex').Knex } knex Database client for persisting grants.
+ * @param { import('@aws-sdk/client-sqs').SQSClient } sqs AWS SDK client used to issue commands to the SQS API.
+ * @param { string } queueUrl The URL identifying the queue to poll.
+ * @param { Array[import('@aws-sdk/client-sqs').Message] } messages Messages to process from SQS.
+ */
+async function processMessages(knex, sqs, queueUrl, messages) {
+    let grantParseErrorCount = 0;
+    let grantSaveSuccessCount = 0;
+    let grantSaveErrorCount = 0;
+
+    return Promise.all(messages.map(async (message) => {
+        console.log('Processing message:', message.Body);
+
+        let grant;
+        try {
+            grant = sqsMessageToGrant(message.Body);
+        } catch (e) {
+            grantParseErrorCount += 1;
+            console.error('Error parsing grant from SQS message:', e);
+            return;
+        }
+
+        try {
+            await upsertGrant(knex, grant);
+            grantSaveSuccessCount += 1;
+        } catch (e) {
+            grantSaveErrorCount += 1;
+            console.error(`Error on insert/update row with grant_id ${grant.grant_id}:`, e);
+            return;
+        }
+
+        try {
+            await deleteMessage(sqs, queueUrl, message.ReceiptHandle);
+        } catch (e) {
+            console.log(
+                `Error deleting SQS message with receipt handle ${message.ReceiptHandle} `,
+                `for grant ${grant.grant_id}`,
+            );
+            throw e;
```

Comment:
When this is thrown, will other async processes be cancelled as well?

Reply:
Yes. I think that's okay in this case for two reasons:
1. If one of the messages fails to delete, it's pretty likely that all of the others will also (due to a bug in the code, missing permission, a connectivity error, or some other issue where the impact isn't limited to one single message).
2. In the worst-case scenario, where:
    - 10 messages were received.
    - 9 of the messages were just saved as grants to the database but haven't been deleted.
    - 1 of the loops was ahead of the others and encountered a completely one-off error when deleting its message.
    
    Then all 10 messages are returned to the queue and will eventually be consumed again. However, their data will have already been saved to the db, so this time around, nothing will actually be modified, and then they'll (presumably) be successfully deleted. A bit of unnecessary work would have occurred, but I think the per-batch processing costs are fairly negligible. 

That said, we'll want to make sure we can monitor these (and other) failure scenarios. It just occurred to me that the Datadog instrumentation recently added to the Express API parts of this repo won't actually integrate with this script, so I'll go ahead and add that now.